src/org/traccar/protocol/TeltonikaProtocolDecoder.java

```diff
@@ -240,27 +248,54 @@ private void decodeLocation(Position position, ChannelBuffer buf, int codec) {
         }
 
         if (channel != null) {
-            ChannelBuffer response = ChannelBuffers.directBuffer(4);
-            response.writeInt(count);
-            channel.write(response);
+            if (channel instanceof DatagramChannel) {
+                ChannelBuffer response = ChannelBuffers.directBuffer(5);
+                response.writeShort(3);
+                response.writeShort(packetId);
+                response.writeByte(0x02);
+                channel.write(response, remoteAddress);
+            } else {
+                ChannelBuffer response = ChannelBuffers.directBuffer(4);
+                response.writeInt(count);
+                channel.write(response);
+            }
         }
 
         return positions;
     }
 
     @Override
-    protected Object decode(
-            Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+    protected Object decode(Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
 
         ChannelBuffer buf = (ChannelBuffer) msg;
 
+        if (channel instanceof DatagramChannel) {
+            return decodeUdp(channel, remoteAddress, buf);
+        } else {
+            return decodeTcp(channel, remoteAddress, buf);
+        }
+    }
+
+    protected Object decodeTcp(Channel channel, SocketAddress remoteAddress, ChannelBuffer buf) throws Exception {
+
         if (buf.getUnsignedShort(0) > 0) {
             parseIdentification(channel, remoteAddress, buf);
         } else {
-            return parseData(channel, remoteAddress, buf);
+            return parseData(channel, remoteAddress, buf, 4, 0);
         }
 
         return null;
     }
 
+    protected Object decodeUdp(Channel channel, SocketAddress remoteAddress, ChannelBuffer buf) throws Exception {
+
+        int packetId = buf.getShort(2);
```

Comment:
Why are you using `getXxx` methods instead of `readXxx`? I think it would simplify things if you just use read. And you won't need any offsets.